src/core/Akka.Streams/Dsl/Valve.cs

```diff
@@ -0,0 +1,181 @@
+    {
+        #region Logic
+
+        private sealed class ValveGraphStageLogic : GraphStageLogic
+        {
+            private readonly Valve<T> _valve;
+            private readonly TaskCompletionSource<IValveSwitch> _completion;
+            private SwitchMode _mode;
+            private readonly IValveSwitch _switch;
+
+            private bool IsOpen => _mode == SwitchMode.Open;
+
+            public ValveGraphStageLogic(Valve<T> valve, TaskCompletionSource<IValveSwitch> completion) : base(valve.Shape)
+            {
+                _valve = valve;
+                _completion = completion;
+                _mode = valve._mode;
+
+                var flipCallback = GetAsyncCallback<Tuple<SwitchMode, TaskCompletionSource<bool>>>(FlipHandler);
+                var getModeCallback = GetAsyncCallback<TaskCompletionSource<SwitchMode>>(t => t.SetResult(_mode));
+
+                _switch = new ValveSwitch(flipCallback, getModeCallback);
+
+                SetHandler(_valve.In, onPush: () =>
+                {
+                    if (IsOpen)
+                        Push(_valve.Out, Grab(_valve.In));
+                });
+
+                SetHandler(_valve.Out, onPull: () =>
+                {
+                    if (IsOpen)
+                        Pull(_valve.In);
+                });
+            }
+
+            public override void PreStart()
+            {
+                _completion.SetResult(_switch);
+            }
+
+            void FlipHandler(Tuple<SwitchMode, TaskCompletionSource<bool>> t)
+            {
+                var flipToMode = t.Item1;
+                var completion = t.Item2;
+
+                var succeed = false;
+
+                if (flipToMode != _mode)
+                {
+                    if (_mode == SwitchMode.Close)
+                    {
+                        if (IsAvailable(_valve.In))
+                            Push(_valve.Out, Grab(_valve.In));
+                        else if (IsAvailable(_valve.Out) && !HasBeenPulled(_valve.In))
+                            Pull(_valve.In);
+
+                        _mode = SwitchMode.Open;
+                    }
+                    else
+                        _mode = SwitchMode.Close;
+
+                    succeed = true;
+                }
+
+                completion.SetResult(succeed);
+            }
+        }
+
+        #endregion
+
+        private readonly SwitchMode _mode;
+
+        /// <summary>
+        /// Creates already open <see cref="Valve{T}"/>
+        /// </summary>
+        public Valve() : this(SwitchMode.Open)
+        {
+        }
+
+        /// <summary>
+        /// Creates <see cref="Valve{T}"/> with inital <paramref name="mode"/>
+        /// </summary>
+        /// <param name="mode">state of the valve at the startup of the flow</param>
+        public Valve(SwitchMode mode)
+        {
+            _mode = mode;
+
+            In = new Inlet<T>("valve.in");
+            Out = new Outlet<T>("valve.out");
```

Comment:
could be inlined with the property definition if you want to save two lines ;)

Reply:
I followed "Retry.cs" code-style here - inlets and outlets are initialized in constructor, not in property declaration